File: vm/src/main/java/com/libreshockwave/vm/trace/InstructionAnnotator.java
package com.libreshockwave.vm.trace;

import com.libreshockwave.chunks.ScriptChunk;
import com.libreshockwave.lingo.Opcode;

import java.util.List;

/**
 * Builds human-readable annotations for Lingo bytecode instructions.
 * Shared utility used by both the debugger UI and tracing system.
 */
public final class InstructionAnnotator {

    private InstructionAnnotator() {
        // Utility class
    }

    /**
     * Build annotation string for an instruction.
     *
     * @param script       the script containing the instruction
     * @param handler      the handler containing the instruction (used for local/param name resolution)
     * @param instr        the instruction to annotate
     * @param resolveNames if true, resolve local/param variable names from handler metadata
     * @return human-readable annotation string, or empty string if no annotation applies
     */
    public static String annotate(ScriptChunk script, ScriptChunk.Handler handler,
                                   ScriptChunk.Handler.Instruction instr, boolean resolveNames) {
        Opcode op = instr.opcode();
        int arg = instr.argument();

        return switch (op) {
            case PUSH_INT8, PUSH_INT16, PUSH_INT32 -> "<" + arg + ">";
            case PUSH_FLOAT32 -> "<" + Float.intBitsToFloat(arg) + ">";
            case PUSH_CONS -> {
                List<ScriptChunk.LiteralEntry> literals = script.literals();
                if (arg >= 0 && arg < literals.size()) {
                    yield "<" + literals.get(arg).value() + ">";
                }
                yield "<literal#" + arg + ">";
            }
            case PUSH_SYMB -> "<#" + script.resolveName(arg) + ">";
            case GET_LOCAL, SET_LOCAL -> {
                if (resolveNames && handler != null && arg >= 0 && arg < handler.localNameIds().size()) {
                    yield "<" + script.resolveName(handler.localNameIds().get(arg)) + ">";
                }
                yield "<local" + arg + ">";
            }
            case GET_PARAM -> {
                if (resolveNames && handler != null && arg >= 0 && arg < handler.argNameIds().size()) {
                    yield "<" + script.resolveName(handler.argNameIds().get(arg)) + ">";
                }
                yield "<param" + arg + ">";
            }
            case GET_GLOBAL, SET_GLOBAL, GET_GLOBAL2, SET_GLOBAL2 -> "<" + script.resolveName(arg) + ">";
            case GET_PROP, SET_PROP -> "<me." + script.resolveName(arg) + ">";
            case LOCAL_CALL -> {
                var handlers = script.handlers();
                if (arg >= 0 && arg < handlers.size()) {
                    yield "<" + script.getHandlerName(handlers.get(arg)) + "()>";
                }
                yield "<handler#" + arg + "()>";
            }
            case EXT_CALL, OBJ_CALL -> "<" + script.resolveName(arg) + "()>";
            case JMP, JMP_IF_Z -> "<offset " + arg + " -> " + (instr.offset() + arg) + ">";
            case END_REPEAT -> "<back " + arg + " -> " + (instr.offset() - arg) + ">";
            default -> "";
        };
    }

    /**
     * Build annotation string without local/param name resolution.
     * Convenience method for tracing where handler context may not be available.
     *
     * @param script the script containing the instruction
     * @param instr  the instruction to annotate
     * @return human-readable annotation string
     */
    public static String annotate(ScriptChunk script, ScriptChunk.Handler.Instruction instr) {
        return annotate(script, null, instr, false);
    }
}

File: cpp/src/lingo/vm/trace/InstructionAnnotator.cpp
#include "libreshockwave/lingo/vm/trace/InstructionAnnotator.hpp"

#include <bit>
#include <cstdint>
#include <sstream>
#include <string>
#include <variant>

namespace libreshockwave::lingo::vm::trace {
namespace {

std::string literalValueToAnnotation(const chunks::ScriptChunk::LiteralValue& value) {
    if (const auto* stringValue = std::get_if<std::string>(&value)) {
        return *stringValue;
    }
    if (const auto* intValue = std::get_if<int>(&value)) {
        return std::to_string(*intValue);
    }
    if (const auto* bytes = std::get_if<std::vector<std::uint8_t>>(&value)) {
        return "[bytes:" + std::to_string(bytes->size()) + "]";
    }
    return "null";
}

std::string floatArgumentToString(int argument) {
    std::ostringstream out;
    out << std::bit_cast<float>(argument);
    return out.str();
}

} // namespace

std::string InstructionAnnotator::annotate(const chunks::ScriptChunk& script,
                                           const chunks::ScriptChunk::Handler* handler,
                                           const chunks::ScriptChunk::Instruction& instruction,
                                           const chunks::ScriptNamesChunk* names,
                                           bool resolveNames) {
    const int arg = instruction.argument;
    switch (instruction.opcode) {
        case Opcode::PUSH_INT8:
        case Opcode::PUSH_INT16:
        case Opcode::PUSH_INT32:
            return "<" + std::to_string(arg) + ">";
        case Opcode::PUSH_FLOAT32:
            return "<" + floatArgumentToString(arg) + ">";
        case Opcode::PUSH_CONS: {
            const auto& literals = script.literals();
            if (arg >= 0 && arg < static_cast<int>(literals.size())) {
                return "<" + literalValueToAnnotation(literals[static_cast<std::size_t>(arg)].value) + ">";
            }
            return "<literal#" + std::to_string(arg) + ">";
        }
        case Opcode::PUSH_SYMB:
            return "<#" + script.resolveName(arg, names) + ">";
        case Opcode::GET_LOCAL:
        case Opcode::SET_LOCAL:
            if (resolveNames && handler != nullptr && arg >= 0 &&
                arg < static_cast<int>(handler->localNameIds.size())) {
                return "<" + script.resolveName(handler->localNameIds[static_cast<std::size_t>(arg)], names) + ">";
            }
            return "<local" + std::to_string(arg) + ">";
        case Opcode::GET_PARAM:
            if (resolveNames && handler != nullptr && arg >= 0 &&
                arg < static_cast<int>(handler->argNameIds.size())) {
                return "<" + script.resolveName(handler->argNameIds[static_cast<std::size_t>(arg)], names) + ">";
            }
            return "<param" + std::to_string(arg) + ">";
        case Opcode::GET_GLOBAL:
        case Opcode::SET_GLOBAL:
        case Opcode::GET_GLOBAL2:
        case Opcode::SET_GLOBAL2:
            return "<" + script.resolveName(arg, names) + ">";
        case Opcode::GET_PROP:
        case Opcode::SET_PROP:
            return "<me." + script.resolveName(arg, names) + ">";
        case Opcode::LOCAL_CALL: {
            const auto& handlers = script.handlers();
            if (arg >= 0 && arg < static_cast<int>(handlers.size())) {
                return "<" + script.getHandlerName(handlers[static_cast<std::size_t>(arg)], names) + "()>";
            }
            return "<handler#" + std::to_string(arg) + "()>";
        }
        case Opcode::EXT_CALL:
        case Opcode::OBJ_CALL:
            return "<" + script.resolveName(arg, names) + "()>";
        case Opcode::JMP:
        case Opcode::JMP_IF_Z:
            return "<offset " + std::to_string(arg) + " -> " +
                   std::to_string(instruction.offset + arg) + ">";
        case Opcode::END_REPEAT:
            return "<back " + std::to_string(arg) + " -> " +
                   std::to_string(instruction.offset - arg) + ">";
        default:
            return "";
    }
}

std::string InstructionAnnotator::annotate(const chunks::ScriptChunk& script,
                                           const chunks::ScriptChunk::Instruction& instruction,
                                           const chunks::ScriptNamesChunk* names) {
    return annotate(script, nullptr, instruction, names, false);
}

} // namespace libreshockwave::lingo::vm::trace
